File: lab6/6-1.py
import random


def quick_select(nums, left, right, k):
    if left == right:
        return nums[left]

    pivot_index = random.randint(left, right)
    pivot_index = partition(nums, left, right, pivot_index)

    # 计算pivot位置在排序后数组中的索引（从0开始）
    target_index = len(nums) - k  # 第k大元素在排序数组中的位置（从小到大排序）

    if pivot_index == target_index:
        return nums[pivot_index]
    elif pivot_index > target_index:  # 目标在左侧
        return quick_select(nums, left, pivot_index - 1, k)
    else:  # 目标在右侧
        return quick_select(nums, pivot_index + 1, right, k)


def partition(nums, left, right, pivot_index):
    pivot_value = nums[pivot_index]
    # 将pivot元素移到右侧
    nums[pivot_index], nums[right] = nums[right], nums[pivot_index]
    store_index = left

    # 这里使用 < 而不是 >，因为我们将按照升序排序来考虑索引
    for i in range(left, right):
        if nums[i] < pivot_value:
            nums[store_index], nums[i] = nums[i], nums[store_index]
            store_index += 1

    # 把pivot元素放回正确的位置
    nums[right], nums[store_index] = nums[store_index], nums[right]
    return store_index


def find_kth_largest(nums, k):
    return quick_select(nums, 0, len(nums) - 1, k)


# 输入
n = int(input())  # 数组的长度
nums = list(map(int, input().split()))  # 数组
k = int(input())  # k值

# 输出第k个最大的元素
print(find_kth_largest(nums, k))

File: lab6/6-1.cpp
# include <iostream>
# include <vector>
# include <algorithm>
using namespace std;
//快速选择算法，用于找到数组中第K个最大的元素
int quickSelect(vector<int>&nums,int left,int right,int k) {
    //选择数组中的最后一个元素作为基准
    int pivot =nums[right];
    int i=left;
    //将小于pivot的元素移到左边，大于pivot的元素移到右边
    for(int j=left;j<right;j++) {
        if(nums[j]>pivot) {
            swap(nums[i],nums[j]);
            i++;
        }
    }
    swap(nums[right],nums[i]);//将基准放到正确的位置
    //如果i的位置正好是 第K个大的元素，返回
    if(i==k) {
        return nums[i];
    }
    //如果i的位置大于k，说明第K大的元素在左边部分
    else if (i>k) {
        return quickSelect(nums,left,i-1,k);
    }
    //如果i的位置小于k，说明第K大的元素在右边部分
    else {
        return quickSelect(nums,i+1,right,k);
    }
}
int main() {
    int n,k;
    cin>>n;
    vector<int>nums(n);
    for(int i=0;i<n;i++) {
        cin>>nums[i];
    }
    cin>>k;//输入k值
    //调用快速选择算法,k的值是从0开始的，所以需要传递k-1
    cout<<quickSelect(nums,0,n-1,k-1)<<endl;
    return 0;
}
